### PR TITLE
[GEOS-7281] move input-less processes to RUNNING state during execution

### PR DESCRIPTION
**Pull request by aaime**
Look ma, an insurmountable problem fixed!

### CODE REVIEW

**Comment by bradh** [comment]
Can't say I fully understand the problem (beyond what is in the JIRA ticket: https://osgeo-org.atlassian.net/browse/GEOS-7281), but it looks like a pretty elegant solution.

**Comment by aaime** [comment]
Thanks @bradh, I should have addressed your feedback

**Comment by bradh** [comment]
:+1: 
Recommend merge.

File: src/extension/wps/wps-core/src/main/java/org/geoserver/wps/executor/ProcessStartupFilter.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+        @Override
+        public Map<String, Object> execute(Map<String, Object> input, ProgressListener monitor)
+                throws ProcessException {
+            if(monitor != null) {
+                monitor.started();
+                monitor = new DelegateProgressListener(monitor) {
+                    @Override
+                    public void started() {
+                        // do not pass over, we already called "started"
```
Comment (jodygarnett) [inline src/extension/wps/wps-core/src/main/java/org/geoserver/wps/executor/ProcessStartupFilter.java]: Okay so started is called *for sure*, even if the delegate does not other to use the monitor. This seems a complicated way to introduce a change to how process progress is recorded.

Is there any reason we should not use this approach *all* the time? Like have the engine call *started* and pass the monitor into the execute function already started.

File: src/extension/wps/wps-core/src/main/java/org/geoserver/wps/executor/ProcessStartupFilter.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    class ProcessStartupFactory extends DelegatingProcessFactory {
+
+        public ProcessStartupFactory(ProcessFactory delegate) {
+            super(delegate);
+        }
+        
+        @Override
+        public Process create(Name name) {
+            Process process = delegate.create(name);
+            return new ProcessStartupWrapper(process);
```
Comment (jodygarnett) [inline src/extension/wps/wps-core/src/main/java/org/geoserver/wps/executor/ProcessStartupFilter.java]: So the ProcessStartupWrapper is used in all cases, not only for no argument processes.

Is this problem due to no argument processes, or just a problem with process implementations that do not honor the API contract (by calling started() on the provided monitor).

File: src/extension/wps/wps-core/src/main/java/org/geoserver/wps/executor/ProcessStartupFilter.java (hunk 1 above)
Comment (aaime) [inline src/extension/wps/wps-core/src/main/java/org/geoserver/wps/executor/ProcessStartupFilter.java]: @jodygarnett I don't understand what you're suggesting, could you be more specific and show where you would place the "started" call?
For example, at first I placed it in the DefaultProcessMananger, which was simpler, but that would not have worked for custom process managers.

File: src/extension/wps/wps-core/src/main/java/org/geoserver/wps/executor/ProcessStartupFilter.java (hunk 2 above)
Comment (aaime) [inline src/extension/wps/wps-core/src/main/java/org/geoserver/wps/executor/ProcessStartupFilter.java]: By any process implementation that does not care to check its inputs (at the low level, the API taking a Map).. Could have limited it to no-args processes, but why bother?

File: src/extension/wps/wps-core/src/main/java/org/geoserver/wps/executor/ProcessStartupFilter.java (hunk 1 above)
Comment (jodygarnett) [inline src/extension/wps/wps-core/src/main/java/org/geoserver/wps/executor/ProcessStartupFilter.java]: Chatting here with Dave and am feeling stupid for not noticing it earlier.

The Process execute method contract asks that the monitor be used (my assumption is started, progress, progress, ... done notifications).

The reflection based process implementation is willing to pass in the monitor for developers to meet this API contract (if the last argument is a monitor). For a no-arguments process (or a process that does not accept a monitor argument) I suspect I have a GeoTools bug where "the monitor" is not being correct notified on execute start and execute completion. 

I *like* your delegate approach (of intercepting the started method) and would like to steal it for a GeoTools fix.

Comment (aaime) [inline src/extension/wps/wps-core/src/main/java/org/geoserver/wps/executor/ProcessStartupFilter.java]: Steal as you see fit :-)

**Pull request merged**